torch/csrc/jit/interpreter.cpp

```diff
@@ -130,39 +399,45 @@ Operation createPythonOperation(PythonOp* op) {
         addOutput(entry);
       }
     }
-    builder.writeTo(outputs);
+    builder.writeTo(stack);
+    return 0;
   };
 }
 
 Operation createCppOperation(CppOp* op) {
   std::shared_ptr<autograd::Function> func = op->fn;
   bool has_handle = hasHandleOutput(op);
-  return [=](const list_of_retainable & inputs, list_of_retainable & outputs) {
+  auto num_inputs = op->inputs().size();
+  return [=](Stack & stack) {
     HandleBuilder builder(has_handle);
     autograd::variable_list v_inputs;
-    for(size_t i = 0; i < inputs.size(); i++) {
-      v_inputs.push_back(builder.addInput(inputs[i], op->var_flags[i]));
+    for(size_t i = 0; i < num_inputs; i++) {
+      v_inputs.push_back(builder.addInput(std::move(fromLast(stack, num_inputs - i)), op->var_flags[i]));
     }
+    drop(stack, num_inputs);
     autograd::variable_list v_outputs = (*func)(v_inputs);
     for(auto & output : v_outputs) {
-      outputs.push_back(builder.addOutput(output));
+      stack.push_back(builder.addOutput(output));
     }
-    builder.writeTo(outputs);
+    builder.writeTo(stack);
+    return 0;
   };
 }
 
 Operation createEvalOperation(CppOp * op) {
   bool has_handle_output = hasHandleOutput(op);
-  return [=](const list_of_retainable & inputs,
-             list_of_retainable & outputs) {
-    AutogradHandle * handle_in = dynamic_cast<AutogradHandle*>(inputs.back());
+  auto num_inputs = op->inputs().size();
+  return [=](Stack & stack) {
+    at::Tensor handle_t = std::move(stack.back());
+    AutogradHandle * handle_in = dynamic_cast<AutogradHandle*>(handle_t.get());
     JIT_ASSERT(handle_in);
     HandleBuilder builder(has_handle_output);
     auto& engine = autograd::python::PythonEngine::getDefaultEngine();
     autograd::variable_list v_inputs;
-    for(size_t i = 0; i < inputs.size() - 1; i++) {
-      v_inputs.push_back(builder.addInput(inputs[i], op->var_flags[i]));
+    for(size_t i = 0; i < num_inputs - 1; i++) {
+      v_inputs.push_back(builder.addInput(std::move(fromLast(stack, num_inputs - i)), op->var_flags[i]));
     }
+    drop(stack, num_inputs);
```

Comment:
It would be nice to never have to use `drop` only after we move all tensors out of the stack. Otherwise it might become a bit messed up in case anything throws before we reach this line

Reply:
Not sure I understand this comment. If something throws then the interpreter itself will get destructed and it will release the entire stack, but I don't think that is what you asked about.